src/Phan/AST/ContextNode.php

```
@@ -2203,8 +2203,21 @@ public function getClassConstListInternal(bool $should_return_first_match): arra
         }
 
         $constant_name = $node->children['const'];
+        if (\PHP_VERSION_ID >= 80300) {
+            if ($constant_name instanceof Node) {
+                $constant_name = UnionTypeVisitor::anyStringLiteralForNode($this->code_base, $this->context, $constant_name);
+            }
+        }
         if (!is_string($constant_name)) {
```

This branch can also be entered when phan is unable to infer a literal for the property, e.g. in the following code (that I'd suggest adding as a test case):

```
$constName = 'MY_CONST_' . rand();
echo C5::{$constName};
```

This would always emit an issue. Instead:
- If we're in PHP < 8.3 and `$constant_name` is not a string (can be either a scalar like `int`, or Node), throw an IssueException immediately as the syntax is invalid
- Else, in PHP 8.3+, if the resolved constant name isn't a string, we should throw NodeException instead, because it's us who can't parse the node, but the code is probably valid.

A further improvement (that can maybe be done as a follow-up) would be to be even more specific when throwing the exception in the PHP 8.3 case, by reporting invalid constant names to the user. Basically the same as `throwExceptionForInvalidPropertyName`: if we know the name node can't be a string (because it's a non-string literal, or a Node whose UnionType can't cast to string), it's a user error (IssueException) and not just an internal error to silently ignore (NodeException).

---

Thank you for your review!

I have revised the source code based on your review.
I implemented error reporting based on `throwExceptionForInvalidPropertyName`.
Are the messages and types correct?

I also added code to trigger a NodeException in the test cases.
However, I was unable to confirm this in file-based testing.
I would appreciate it if you could confirm that the current implementation is working as intended.

Thank you so much!

---

Thank you! The implementation seems correct. I don't think it's possible to see the effect of a NodeException in tests, as callers just ignore it. I left inline comments for some error message suggestions!